src/views/RequiredSignoffs/ViewSignoff/index.jsx

```diff
@@ -29,106 +39,250 @@ const useStyles = makeStyles(theme => ({
+    left: 0,
+    right: 0,
+  },
 }));
 
-function ViewSignoff(props) {
+function ViewSignoff({ isNewSignoff, ...props }) {
+  const emptyRole = ['', '', { isAdditionalRole: true, id: Math.random() }];
+  const { product, channel } = props.match.params;
   const classes = useStyles();
-  // eslint-disable-next-line no-unused-vars
-  const [products, setProducts] = useState();
-  // eslint-disable-next-line no-unused-vars
-  const [requiredSignoffs, setRequiredSignoffs] = useState(null);
-  const [error, setError] = useState();
-  const [type, setType] = useState('channel');
+  const [channelTextValue, setChannelTextValue] = useState(channel || '');
+  const [productTextValue, setProductTextValue] = useState(product || '');
+  const [type, setType] = useState(channel ? 'channel' : 'permission');
+  const [roles, setRoles] = useState([]);
+  const [additionalRoles, setAdditionalRoles] = useState(
+    isNewSignoff ? [emptyRole] : []
+  );
+  const [requiredSignoffs, getRS] = useAction(getRequiredSignoffs);
+  const [products, getProds] = useAction(getProducts);
+  const isLoading = requiredSignoffs.loading || products.loading;
   const handleTypeChange = ({ target: { value } }) => setType(value);
+  const handleChannelChange = ({ target: { value } }) =>
+    setChannelTextValue(value);
+  const handleProductChange = value => setProductTextValue(value);
+  const handleRoleValueChange = (role, index) => ({ floatValue: value }) => {
+    const setRole = (entry, i) =>
+      i === index ? [entry[0], value, entry[2]] : entry;
+
+    return role[2].isAdditionalRole
+      ? setAdditionalRoles(additionalRoles.map(setRole))
+      : setRoles(roles.map(setRole));
+  };
+
+  const handleRoleNameChange = (role, index) => ({ target: { value } }) => {
+    const result = additionalRoles.map((entry, i) =>
+      i === index ? [value, entry[1], entry[2]] : entry
+    );
+
+    setAdditionalRoles(result);
+  };
+
+  const handleRoleAdd = () => {
+    setAdditionalRoles(additionalRoles.concat([emptyRole]));
+  };
+
+  const handleRoleDelete = (role, index) => () => {
+    const excludeRole = (entry, i) => !(i === index);
+
+    return role[2].isAdditionalRole
+      ? setAdditionalRoles(additionalRoles.filter(excludeRole))
+      : setRoles(roles.filter(excludeRole));
+  };
+
+  // TODO: Add save logic
+  const handleSignoffSave = () => {};
+  // TODO: Add delete logic
+  const handleSignoffDelete = () => {};
 
   useEffect(() => {
-    tryCatch(getProducts()).then(([error, result]) => {
-      if (error) {
-        setError(error);
+    if (isNewSignoff) {
+      return getProds();
+    }
 
-        return;
-      }
+    // eslint-disable-next-line no-unused-vars
+    Promise.all([getProds(), getRS()]).then(([prods, rs]) => {
+      const roles = getRolesFromRequiredSignoffs(rs.data);
 
-      setProducts(result.data.product);
+      setRoles(roles);
     });
+  }, [product, channel]);
+
+  const renderRole = (role, index) => (
+    <Grid
+      key={role[2].id}
```

Comment:
It would be nice if the endpoint returned roles with a unique property like an `id` or perhaps a combination of properties that makes an entry unique. The `key` property [is important](https://reactjs.org/docs/lists-and-keys.html#keys) when mapping over a list. Right now `role[2].id` is being generated with `Math.random()`.

Reply:
`role` itself should be unique for any given product/channel or product permissions (it's part of the primary key in the table it's stored in). That should make it OK here, unless I'm missing something?

Reply:
Erm, clearly that was the wrong account I commented with. It was me, though!

Reply:
By `role`, do you mean the combination of all the properties of a role is what's unique?

Reply:
Yeah, sorry - if we're talking about a response with a list of required signoffs, eg:
```
{"channel":"esr","data_version":1,"product":"SystemAddons","role":"relman","signoffs_required":1},
{"channel":"esr","data_version":1,"product":"SystemAddons","role":"sysaddon","signoffs_required":1},{"channel":"release","data_version":1,"product":"SystemAddons","role":"relman","signoffs_required":1},
```

...product+channel+role is unique.

For permissions, product+role is unique.

Reply:
Awesome. I will codify that right now :)